### PR TITLE
contracts-bedrock: port custom gas token to portal2

### PR DESCRIPTION
**Pull request by tynes**
**Description**

Ports the custom gas token feature to `OptimismPortal2`.
This will enable fault proofs to run on custom gas token
chains.

### CODE REVIEW

**Comment by tynes** [comment]
@coderabbitai summary

**Comment by tynes** [comment]
@coderabbitai Double check that all of the tests from `OptimismPortal.t.sol` involving custom gas token have been ported to `OptimismPortal2.t.sol`

File: packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:3]: @coderabbitai Are all of the tests added in PR https://github.com/ethereum-optimism/optimism/pull/10143 also included in this file?

**Comment by tynes** [comment]
@mds1 - I went through and made sure all of the tests were ported from the original implementation

File: packages/contracts-bedrock/snapshots/storageLayout/OptimismPortal2.json
Comment (mds1) [inline packages/contracts-bedrock/snapshots/storageLayout/OptimismPortal2.json:127]: This storage layout diff matches what I'd expect ✅

File: packages/contracts-bedrock/snapshots/abi/OptimismPortal2.json
Comment (mds1) [inline packages/contracts-bedrock/snapshots/abi/OptimismPortal2.json:24]: Verified the ABI diff as well. Aside from just looking at the ABI diff in this PR I also compared the OptimismPortal and OptimismPortal2 ABIs—no missing or extra gas token methods found ✅

File: packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol
Comment (mds1) [inline packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:44]: Good idea, haven't seen this pattern before

**Review by mds1 — COMMENTED** [review]
This looks great, seems like the merge was pretty straightforward as expected. 

I assume in a follow up PRs we will delete the original portal then rename portal2 to portal?

>  I went through and made sure all of the tests were ported from the original implementation

Unfortunately I did not get a chance to double check this, but would just ask if we can get someone else to verify none are missing before merge. Worst case I can confirm once I'm back from vacation

**Review by mds1 — APPROVED** [review]

**Comment by tynes** [comment]
> I assume in a follow up PRs we will delete the original portal then rename portal2 to portal?

Yup, exactly!

> Unfortunately I did not get a chance to double check this, but would just ask if we can get someone else to verify none are missing before merge. Worst case I can confirm once I'm back from vacation

Nobody has taken the time to do this review yet, so this is still pending. No major rush, this is nice to have to enable the interop devnet to run with the fault proof code, will rebase https://github.com/ethereum-optimism/optimism/pull/11051 on top of this after its merged on `develop`

File: packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol
Comment (tynes) [reply packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:44]: Thanks, we will want to delete this in the future, its just a sanity check to guarantee that we don't use the wrong contracts in any test

**Comment by tynes** [comment]
@mds1 There has been no progress on this PR while you have been on vacation, but I did go through myself to double check that all the tests were ported and didn't find anything missing

**Comment by tynes** [comment]
@mds1 bump on this, I know you are busy but I do think this is good to merge. It will unblock me from doing another custom gas token release

**Review by mds1 — APPROVED** [review]
> There has been no progress on this PR while you have been on vacation, but I did go through myself to double check that all the tests were ported and didn't find anything missing

Thanks, I also confirmed with a regex search of `\bfunction\s+test.*Gas.*\b` to find all gas related tests in `OptimismPortal.t.sol` and confirm it matches the set in `OptimismPortal2.t.sol`, which it did

Since no other changes have been made since my last review, this LGTM. Nice work, it's great to have these merged

**Pull request merged**
